feat(stdlib/strings): add length and substring functionality includin…

ewendai opened the pull request:
…g unit and integration tests


### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

nathanielc commented on stdlib/strings/strings.go:
```diff
@@ -382,7 +384,69 @@ func generateUnicodeIsFunction(name string, Fn func(rune) bool) values.Function
+}
+
+func generateSubstring(name string) values.Function {
+	return values.NewFunction(
```
As a small suggestion I would change these to be `var`s instead of functions as the function is only called once to produce a single value. 

```
var substring = values.NewFunction(...)
```

Then in the init function it becomes 
```

	flux.RegisterPackageValue("strings", "substring", substring)
```

nathanielc commented on stdlib/strings/strings.go:
```diff
@@ -382,7 +384,69 @@ func generateUnicodeIsFunction(name string, Fn func(rune) bool) values.Function
+			}
+
+			if v.Type().Nature() == semantic.String {
+				arr := strings.Split(v.Str(), "")
```
This method is going to be the fastest https://golang.org/pkg/unicode/utf8/#RuneCountInString we should use it.

aanthony1243 commented on stdlib/strings/strings_test.go:
```diff
@@ -1250,3 +1251,154 @@ func TestJoinStr(t *testing.T) {
+			res := result.Str()
+
+			if res != (tc.want) {
+				t.Errorf("string function result %s expected: %v, got: %s", tc.name, tc.want, result)
```
great tests!

nathanielc commented on stdlib/strings/strings.go:
```diff
@@ -382,7 +385,77 @@ func generateUnicodeIsFunction(name string, Fn func(rune) bool) values.Function
+			return nil, fmt.Errorf("indices out of bounds")
+		}
+
+		if (v.Type().Nature() == semantic.String) && (a.Type().Nature() == semantic.Int) && (b.Type().Nature() == semantic.Int) {
```
This type check needs to happen before lines 427-429. Otherwise those methods will panic.

nathanielc commented on stdlib/strings/strings_test.go:
```diff
@@ -1250,3 +1251,163 @@ func TestJoinStr(t *testing.T) {
+	}
+}
+
+func BenchmarkSubstring(b *testing.B) {
```
Did you compare this new implementation to the old one with this benchmark? How did it compare?

ewendai replied:
I did, the one that I kept was consistently faster.

nathanielc commented on stdlib/strings/strings.go:
```diff
@@ -382,7 +385,77 @@ func generateUnicodeIsFunction(name string, Fn func(rune) bool) values.Function
+				if count >= bInt {
+					break
+				}
+				newStr = strings.Join([]string{newStr, string(c)}, "")
```
This line is still constructing a new string with each loop which is allocation heavy. You can use Go's substring here to help.

I am thinking something along the lines of finding the byteStart and byteEnd indexes and then doing this after the loop.

```
return values.NewString(vStr[byteStart:byteEnd]), nil
```
Can you try that and use [benchcmp](https://godoc.org/golang.org/x/tools/cmd/benchcmp) to compare the two implementations to see which is better?

ewendai replied:
using `byteStart` and `byteEnd` makes it run about the same or faster, depending on which end of the string I'm getting the substring from. will be changing it to the faster version

nathanielc commented on stdlib/strings/strings.go:
```diff
@@ -382,7 +385,77 @@ func generateUnicodeIsFunction(name string, Fn func(rune) bool) values.Function
+				}
+				count++
+			}
+			return values.NewString(vStr[byteStart:byteEnd]), nil
```
This looks great, can you share the benchmark results? I am interested in seeing how this change performs compared to the previous implementation.

ewendai replied:
How do I share the benchmark results? Just write a line with the previous implementation and time compared to current?

ewendai commented:
**Benchmark Test Summary**
- Slowest Implementation: Split given string using `strings.split`, retrieve the desired parts of the string, return after joining using `strings.joinStr`
- Consistently better: Loop over the given string and add the desired parts of the string together 
- About the same if desired substring is at the beginning but better if desired substring is near the end of the given string: Loop over the given string to find `byteStart` and `byteEnd` and return `givenString[byteStart:byteEnd]`
